### ISSUE TITLE
Replace Boston dataset in examples

### ISSUE THREAD

**Issue by thatlittleboy** [dsgibbons/shap#8]
Test logs are showing 

```
`load_boston` has been removed from scikit-learn since version 1.2.
```

with an elaborate explanation about how the Boston dataset has some ethical issues.

We should find a replacement dataset to use, and change all existing examples to that.

Partially related to #4 .

**Comment by thatlittleboy** [comment]
Look into California housing dataset proposed by https://github.com/slundberg/shap/pull/2873.

**Comment by simonangerbauer** [comment]
As a reference, some occurrences of the boston dataset have already been replaced in slundberg#2501.

**Comment by connortann** [comment]
I think replacing this dataset might not be entirely straightforward. The dataset is made available in SHAP's public API in `shap.datasets.boston` , so removing it from here would be a breaking change for users.

We could tackle this from two angles:

### 1. Fixing the broken code in SHAP's public API

A short-term fix could be to reimplement the `shap.datasets.boston` function by fetching from the original source, as recommended by sklearn:

<details>

```python
import pandas as pd
import numpy as np

data_url = "http://lib.stat.cmu.edu/datasets/boston"
raw_df = pd.read_csv(data_url, sep="\\s+", skiprows=22, header=None)
data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
target = raw_df.values[1::2, 2]
```

</details>

### 2. Deprecating the Boston dataset

There is an argument that the Boston dataset is rather useful for demonstrating AI ethics and bias issues, for example to evaluate fairness metrics with respect to demographics. Shapley values are quite relevant to this domain, so perhaps the dataset should be kept for this purpose.

Alternatively, if it is removed, I think it would be prudent to use a suitable deprecation cycle to avoid users facing unexpected breaking changes. The sklearn maintainers were careful to have a long deprecation period, given that they identified that `load_boston` was a particularly widely-used funtion across tutorials and other packages: https://github.com/scikit-learn/scikit-learn/pull/20729

**Comment by dsgibbons** [comment]
Ah I approved #20 before reading this reply @connortann - sorry about that. I like the deprecation route for now (as used by #20), but am open to rolling back if we would like to discuss this further.

**Comment by thatlittleboy** [comment]
+1 on the deprecation (I don't think there's a strong need to continue using boston dataset).

For posterity, the remaining use of `shap.datasets.boston`:

![image](https://github.com/dsgibbons/shap/assets/30731072/35027b59-29d5-4f46-a55d-416bf4696a7b)

We need to address these before this issue can be closed~

**Comment by dsgibbons** [comment]
Note that for `README.md`, the following code snippets all create plots using `shap.datasets.boston`. All of them need reformatting/fixing so that labels aren't cut off etc. Baseline artifacts from `shap.datasets.boston` can be found [here](https://github.com/dsgibbons/shap/tree/master/docs/artwork)

- `shap.plots.waterfall(shap_values[0])` (needs formatting)
- `shap.plots.force(shap_values[0])` (doesn't plot anything, no obvious error)
- `shap.plots.force(shap_values)` (doesn't plot anything, no obvious error)
- `shap.plots.scatter(shap_values[:,"RM"], color=shap_values)` (needs RM changing to a feature from `shap.datasets.california`)
- `shap.plots.beeswam(shap_values)` (needs formatting)
- `shap.plots.bar(shap_values)` (needs formatting)
